pkg/indexer/reorgHandler.go

```diff
@@ -0,0 +1,156 @@
+package indexer
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/xmtp/xmtpd/pkg/blockchain"
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+)
+
+type ChainReorgHandler interface {
+	FindReorgPoint(detectedAt uint64) (uint64, []byte, error)
+}
+
+type ReorgHandler struct {
+	ctx     context.Context
+	client  blockchain.ChainClient
+	queries *queries.Queries
+}
+
+var (
+	ErrNoBlocksFound = errors.New("no blocks found")
+	ErrGetBlock      = errors.New("failed to get block")
+)
+
+// TODO: Make this configurable?
+const BLOCK_RANGE_SIZE uint64 = 1000
+
+func NewChainReorgHandler(
+	ctx context.Context,
+	client blockchain.ChainClient,
+	queries *queries.Queries,
+) *ReorgHandler {
+	return &ReorgHandler{
+		ctx:     ctx,
+		client:  client,
+		queries: queries,
+	}
+}
+
+// TODO: When reorg range has been calculated, alert clients (TBD)
+func (r *ReorgHandler) FindReorgPoint(detectedAt uint64) (uint64, []byte, error) {
+	startBlock, endBlock := blockRange(detectedAt)
+
+	for {
+		storedBlocks, err := r.queries.GetBlocksInRange(
+			r.ctx,
+			queries.GetBlocksInRangeParams{
+				StartBlock: startBlock,
+				EndBlock:   endBlock,
+			},
+		)
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return 0, nil, fmt.Errorf("failed to get stored blocks: %w", err)
+		}
+
+		if len(storedBlocks) == 0 || errors.Is(err, sql.ErrNoRows) {
+			if startBlock == 0 {
+				return 0, nil, ErrNoBlocksFound
```

Comment:
I worry we are overloading the `ErrNoBlocksFound` error. It seems to mean slightly different things in different places (no blocks in the DB, reorg happened earlier in the chain, empty range)

Reply:
The idea around it is using it when the range starts at 0 and we still didn't find the reorg point, which means the reorg happened since block 0... very unlikely.
Removing it from `searchInRange`, it's true that it doesn't make sense there as we'll be always passing ranges at least length == 1.